gateway/src/node/client_handling/websocket/connection_handler/ecash/mod.rs

```diff
@@ -0,0 +1,481 @@
+                            String::from(
+                                "credential failed to verify on gateway - duplicate payInfo",
+                            ),
+                        ));
+                    }
+                    i -= 1;
+                }
+
+                let mut i = index + 1;
+                while i < inner.len() && inner[i].timestamp() == pay_info.timestamp() {
+                    if inner[i] == pay_info {
+                        return Err(RequestHandlingError::InvalidBandwidthCredential(
+                            String::from(
+                                "credential failed to verify on gateway - duplicate payInfo",
+                            ),
+                        ));
+                    }
+                    i += 1;
+                }
+                Ok(index)
+            }
+        }
+    }
+
+    fn insert_pay_info(&self, pay_info: PayInfo, index: usize) -> Result<(), RequestHandlingError> {
+        let mut inner = self
+            .pay_infos
+            .lock()
+            .map_err(|_| RequestHandlingError::InternalError)?;
+        if index > inner.len() {
+            inner.push(pay_info);
+            return Ok(());
+        }
+        inner.insert(index, pay_info);
+        Ok(())
+    }
+
+    pub async fn check_double_spend(&self, serial_number_bs58: &String) -> bool {
+        if let Some(double_spend_detector) = &self.double_spend_detector {
+            double_spend_detector.check(serial_number_bs58).await
+        } else {
+            false
+        }
+    }
+
+    pub async fn post_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        self.cred_sender
+            .unbounded_send(PendingCredential::new(
+                credential.clone(),
+                self.nyxd_client.read().await.address(),
+                api_clients
+                    .iter()
+                    .map(|client| client.api_client.clone())
+                    .collect(),
+            ))
+            .map_err(|_| RequestHandlingError::InternalError)
+    }
+
+    pub async fn spend_online_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: &CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        let serial_number = credential.data.payment.serial_number_bs58();
+        let proposal_id = match credential.data.typ {
+            CredentialType::Voucher => unimplemented!(),
+            CredentialType::TicketBook => {
+                let res = self
+                    .nyxd_client
+                    .write()
+                    .await
+                    .prepare_credential(serial_number, self.address.to_string(), None)
+                    .await?;
+                let proposal_id = find_attribute(&res.logs, "wasm", BANDWIDTH_PROPOSAL_ID)
```

Comment:
I think I mentioned this before about creating some helper for obtaining the id directly without this nasty parsing code everywhere

Reply:
TO FIX

Reply:
Fixed in eb1c2056a5a36b962e6f1597a29e04a47a59fd51